qcodes/tests/test_autoloadable_channels.py

```diff
@@ -0,0 +1,186 @@
+            r":INST:CHN(\d):HLO": lambda chn: f"Hello from channel {chn}",
+            r":INST:CHN:ADD (\d)": self._channel_catalog.append,
+            r":INST:CHN:DEL (\d)": self._channel_catalog.remove,
+            r":INST:CHN:CAT": lambda: ",".join([str(i) for i in self._channel_catalog])
+        }
+
+
+class SimpleTestChannel(AutoLoadableInstrumentChannel):
+    """
+    A channel to test if we can create and delete channel instances
+    """
+
+    @classmethod
+    def _discover_from_instrument(
+            cls, parent: Instrument, **kwargs) -> List[dict]:
+        """
+        New channels need `name` and `channel` keyword arguments.
+        """
+        channels_str = parent.channel_catalog()
+        kwarg_list = [
+            {"name": f"channel{i}", "channel": int(i)}
+            for i in channels_str.split(",")
+        ]
+
+        return kwarg_list
+
+    @classmethod
+    def get_new_instance_kwargs(cls, parent: Instrument, **kwargs) -> dict:
+        """
+        Find the smallest channel number not yet occupied
+        """
+        channels_str = parent.channel_catalog()
+        existing_channels = [int(i) for i in channels_str.split(",")]
+
+        new_channel = 1
+        while new_channel in existing_channels:
+            new_channel += 1
+
+        kwargs = {
+            "name": f"channel{new_channel}",
+            "channel": new_channel
+        }
+
+        return kwargs
+
+    def __init__(
+            self,
+            parent: Union[Instrument, 'InstrumentChannel'],
+            name: str,
+            channel: int,
+            existence: bool = False,
+            channel_list: 'AutoLoadableChannelList' = None,
+    ) -> None:
+
+        super().__init__(parent, name, existence, channel_list)
+        self._channel = channel
+
+        self.add_parameter(
+            "hello",
+            get_cmd=f":INST:CHN{self._channel}:HLO"
+        )
+
+    def _create(self)->None:
+        """Create the channel on the instrument"""
+        self.parent.write(f":INST:CHN:ADD {self._channel}")
+
+    def _delete(self)->None:
+        """Remove the channel from the instrument"""
+        self.write(f":INST:CHN:DEL {self._channel}")
+
+
+class TestInstrument(Instrument):
+    """
+    This dummy instrument allows the creation and deletion of
+    channels
+    """
+
+    def __init__(self, name):
+        super().__init__(name)
+
+        self._backend = TestInstrumentBackend()
+
+        self.add_parameter(
+            "channel_catalog",
+            get_cmd=":INST:CHN:CAT",
+        )
+
+    def write_raw(self, cmd: str):
+        return self._backend.send(cmd)
+
+    def ask_raw(self, cmd: str):
+        return self._backend.send(cmd)
+
+
+class DummyInstrumentWithChannelList(TestInstrument):
+    def __init__(self, name):
+        super().__init__(name)
+
+        channels = AutoLoadableChannelList(self, "channels", SimpleTestChannel)
+        self.add_submodule("channels", channels)
+
+
+def test_sanity():
```

Comment:
thanks for the test! but an example notebook with at least (for now) bare intentions of this infra, would be appreciated. This will also help to (if necessary) improve the design, strip down excessive code, and perhaps even add something. (yes, yes, i know that everything is expressed in the description of the linked issue, but still :) )

Reply:
I honestly think that adding a driver which uses this infrastructure will be illuminating, perhaps even more so than a notebook as it will show the complexities involved with a real instrument. I will make the PNA driver use the infrastructure which I think will be helpful.